### PR TITLE
[Experiment] feat(react-hooks): add react-hooks

### PR DESCRIPTION
**Pull request by joshblack**
This is an exploratory PR to gauge interest in a dedicated `react-hooks` package that could surface helpful hooks not only for our own project, but to be exposed for others to consume.

In this PR, we introduce three hooks:

- `useId`: equivalent to our `instanceId` helpers, provides a unique id for the lifecycle of a component
- `usePortalNode`: is a way to get access to a node in which one could render a portal, used in `useAnnouncer` to get access to a singleton `#carbon-announcer` node
- `useAnnouncer`: a way to control displaying polite and assertive messages in dedicated aria-live regions

In the future, it would be awesome to consolidate hooks we use across projects to make it easier for product teams to use/leverage them. I would love to expand this package to add support for window events, in particular for things like scroll/resize/scroll direction

#### Changelog

**New**

- `packages/react-hooks`
  - `useId`
  - `usePortalNode`
  - `useAnnouncer`

**Changed**

**Removed**

#### Testing / Reviewing

- Each of the hooks has an associated story and test suite to check out
- `useAnnouncer` is an interesting use case, here's a video for a full overview of what to expect when testing with VO

demo

### CODE REVIEW

**Review by asudoh — APPROVED** [review]

File: packages/react-hooks/src/useAnnouncer.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+/**
+ * Provides an `announce` method that allows a user to queue up an assertive or
+ * polite message to the user. This message is displayed in an `aria-live`
+ * region with the appropriate mode and the message is set as its text content.
+ * This `aria-live` region is the same for all components, so ordering of
+ * messages sent is important.
+ */
+export function useAnnouncer() {
+  const node = usePortalNode('carbon-announcer', node => {
```
Comment (joshblack) [inline packages/react-hooks/src/useAnnouncer.js]: I'm curious if we want a callback to `usePortalNode`, or if we should just put this in an effect 🤔

Comment (dakahn) [reply packages/react-hooks/src/useAnnouncer.js]: this is pretty clear to read. I like it 😸

File: packages/react-hooks/src/useAnnouncer.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    if (!node) {
+      return;
+    }
+
+    // Each time the mode or announcement changes, we'll want to update the
+    // message at that node.
+    const assertiveNode = node.querySelector('#carbon-assertive-announcement');
+    const politeNode = node.querySelector('#carbon-polite-announcement');
+    const timeoutId = setTimeout(() => {
+      if (mode === 'assertive' && assertiveNode.textContent !== announcement) {
```
Comment (dakahn) [inline packages/react-hooks/src/useAnnouncer.js:67]: This is very readable 👍

**Review by dakahn — APPROVED** [review]
This seems great! Should we get a ticket going that tracks some usage documentation around -- i'm thinking in particular the announcer, but the other hooks as well. I'm also wondering if we should get something on paper about what _warrants_ making something a hook.

**Pull request merged**